2D context state flag for inverse transform is not maintained correctly

kkinnunen-apple opened the pull request:
#### e0c0c143f85c8f2bfd2b4e00d0c9aa944ab30902
<pre>
2D context state flag for inverse transform is not maintained correctly
<a href="https://bugs.webkit.org/show_bug.cgi?id=294117">https://bugs.webkit.org/show_bug.cgi?id=294117</a>
<a href="https://rdar.apple.com/149686936">rdar://149686936</a>

Reviewed by Said Abou-Hallawa.

293589@main would make CanvasRenderingContext2DBase::restore() use
    if (hasInvertibleTransform())
instead of:
    if (auto inverse = transform.inverse())

This would cause dereference of unengaged optional for the `inverse`,
because the operations could make the transform non-invertible
even though the operation parameters would not indicate so.
E.g. consider:
  canvas.transform(..., number1, ...);
  canvas.scale(1, bignumber2);

Some pairs of number1, bignumber2 could make the transform
non-invertible.

Fix by modifying the CanvasPath::m_hasInvertibleTransform and
modifiableState().transform only in one function, the added
CanvasRenderingContext2DBase::updateState. Use the actual
inversible state as the source of the m_hasInvertibleTransform.

* LayoutTests/fast/canvas/canvas-state-non-invertible-ctm-expected.txt: Added.
* LayoutTests/fast/canvas/canvas-state-non-invertible-ctm.html: Added.
* Source/WebCore/html/canvas/CanvasRenderingContext2D.cpp:
(WebCore::CanvasRenderingContext2D::drawFocusIfNeededInternal):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::restore):
(WebCore::CanvasRenderingContext2DBase::scale):
(WebCore::CanvasRenderingContext2DBase::rotate):
(WebCore::CanvasRenderingContext2DBase::translate):
(WebCore::CanvasRenderingContext2DBase::transform):
(WebCore::CanvasRenderingContext2DBase::resetTransform):
(WebCore::CanvasRenderingContext2DBase::updateStateTransform):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.h:

Canonical link: <a href="https://commits.webkit.org/296373@main">https://commits.webkit.org/296373@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fba9ca116c58e154c22a2771f5ade379e8e643ed

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108148 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27811 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18231 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113359 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58634 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110111 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28509 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36361 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82113 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111096 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22588 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97424 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62544 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22002 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15560 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58099 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/91952 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15616 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116485 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35214 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/25930 "Found 1 new test failure: fast/webgpu/nocrash/fuzz-291180.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91140 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35588 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93701 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/90935 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23209 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35826 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13587 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/31012 "Build was cancelled. Recent messages:Printed configuration") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35113 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40669 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34847 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38205 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36514 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/8d556239fb0a4750ff418d2bd2e9cbe6c8bc6fc7. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46420

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e8edc82b0196b6d567507a18e03dc970c71baa20. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46420

kkinnunen-apple commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -914,8 +910,8 @@ void CanvasRenderingContext2DBase::translate(double tx, double ty)
-
-    modifiableState().transform = newTransform;
+    if (!updateStateTransform(newTransform)) [[unlikely]]
+        return;
```
Translation might make transformation non-invertible in cases where the addition overflows the double range, producing infinities. Added a comment.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/fba9ca116c58e154c22a2771f5ade379e8e643ed. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46420

kkinnunen-apple committed:
2D context state flag for inverse transform is not maintained correctly
https://bugs.webkit.org/show_bug.cgi?id=294117
rdar://149686936

Reviewed by Said Abou-Hallawa.

293589@main would make CanvasRenderingContext2DBase::restore() use
    if (hasInvertibleTransform())
instead of:
    if (auto inverse = transform.inverse())

This would cause dereference of unengaged optional for the `inverse`,
because the operations could make the transform non-invertible
even though the operation parameters would not indicate so.
E.g. consider:
  canvas.transform(..., number1, ...);
  canvas.scale(1, bignumber2);

Some pairs of number1, bignumber2 could make the transform
non-invertible.

Fix by modifying the CanvasPath::m_hasInvertibleTransform and
modifiableState().transform only in one function, the added
CanvasRenderingContext2DBase::updateState. Use the actual
inversible state as the source of the m_hasInvertibleTransform.

* LayoutTests/fast/canvas/canvas-state-non-invertible-ctm-expected.txt: Added.
* LayoutTests/fast/canvas/canvas-state-non-invertible-ctm.html: Added.
* Source/WebCore/html/canvas/CanvasRenderingContext2D.cpp:
(WebCore::CanvasRenderingContext2D::drawFocusIfNeededInternal):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::restore):
(WebCore::CanvasRenderingContext2DBase::scale):
(WebCore::CanvasRenderingContext2DBase::rotate):
(WebCore::CanvasRenderingContext2DBase::translate):
(WebCore::CanvasRenderingContext2DBase::transform):
(WebCore::CanvasRenderingContext2DBase::resetTransform):
(WebCore::CanvasRenderingContext2DBase::updateStateTransform):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.h:

Canonical link: https://commits.webkit.org/296373@main
